rest_framework/permissions.py

```diff
@@ -51,24 +52,45 @@ class AND:
     def __init__(self, op1, op2):
         self.op1 = op1
         self.op2 = op2
+        self.message1 = getattr(op1, 'message', None)
+        self.message2 = getattr(op2, 'message', None)
+        self.message = None
 
     def has_permission(self, request, view):
-        return (
-            self.op1.has_permission(request, view) and
-            self.op2.has_permission(request, view)
-        )
+        if not self.op1.has_permission(request, view):
+            self.message = self.message1
+            return False
+
+        if not self.op2.has_permission(request, view):
+            self.message = self.message2
+            return False
+
+        return True
 
     def has_object_permission(self, request, view, obj):
-        return (
-            self.op1.has_object_permission(request, view, obj) and
-            self.op2.has_object_permission(request, view, obj)
-        )
+        if not self.op1.has_object_permission(request, view, obj):
+            self.message = self.message1
+            return False
+
+        if not self.op2.has_object_permission(request, view, obj):
+            self.message = self.message2
+            return False
+
+        return True
 
 
 class OR:
     def __init__(self, op1, op2):
         self.op1 = op1
         self.op2 = op2
+        self.message1 = getattr(op1, 'message', None)
+        self.message2 = getattr(op2, 'message', None)
+        self.message = self.message1 or self.message2
+        if self.message1 and self.message2:
+            self.message = _('"{message1}" OR "{message2}"').format(
+                message1=self.message1,
+                message2=self.message2,
+            )
```

Comment:
There's no point in using `_(...)` for `'"{message1}" OR "{message2}"'` string because we don't have the ability to translate this string which is compiled in runtime.

Let's do this:
```python
self.message = '"{message1}" {or} "{message2}"'.format(
    message1=self.message1,
    or=_("or")
    message2=self.message2,
)
```